### PR TITLE
Painless: Restructure Spec Documentation

### PR DESCRIPTION
**Pull request by jdconrad**
This splits general-syntax into the remaining spec sections statements, scripts, functions, lambdas, and regexes.

This also splits the operators into different sections in the Painless spec to reduce the size of the original operators page from enormous to manageable.  Also includes clean up of the general operators section.  Subsequent operators sections will be cleaned up in different PRs.

Note this PR looks huge, but it's actually not.  Most of the lines are the split of operators into different pages.

### CODE REVIEW

File: docs/painless/painless-operators-boolean.asciidoc [hunk 1]
```diff
@@ -0,0 +1,516 @@
+[[painless-operators-boolean]]
+=== Operators: Boolean
+
+[[boolean-not-operator]]
+==== Boolean Not
+
+Boolean not will flip a boolean value from true to false or false to true using the bang operator.  The format is a bang operator followed by an expression.
+
```
Comment (debadair) [inline docs/painless/painless-operators-boolean.asciidoc]: I think I'd put `not` in code font or all caps to make it a bit easier to read correctly. I'd probably say: Boolean `not` flips a boolean value between `true` and `false` and is represented by the ! (bang) operator. The format is the bang operator followed by an expression.

File: docs/painless/painless-operators-boolean.asciidoc [hunk 2]
```diff
@@ -0,0 +1,516 @@
...
+*Grammar:*
+[source,ANTLR4]
+----
+boolean_not: '!' expression;
+----
+
+Note that def types will be assumed to be of the boolean type.  Any def type evaluated at run-time that does not represent a boolean will result in an error.  Non-boolean expressions will result in an error.
+
```
Comment (debadair) [inline docs/painless/painless-operators-boolean.asciidoc]: Generally avoid future tense. Also, runtime doesn't need to be hyphenated. I'd say "Note that def types are assumed to be of type boolean. Any def type evaluated at runtime that does not represent a boolean results in an error."

The last sentence is not clear--is that, "Preceding a non-boolean expression with the  bang operator results in an error."?

File: docs/painless/painless-operators-boolean.asciidoc [hunk 3]
```diff
@@ -0,0 +1,516 @@
...
+Greater than compares two numerical expressions where a resultant boolean value will be true if the left-side expression is a larger value than the right-side expression otherwise false.  The format is an expression, followed by the right angle operator, and a closing expression.
+
+*Grammar:*
+[source,ANTLR4]
+----
+greater_than: expression '>' expression;
+----
+
+A numeric promotion may occur during a greater than operation.  A def type evaluated at run-time will follow the same promotion table at run-time following whatever type def represents. Non-numeric expressions will result in an error.
+
```
Comment (debadair) [inline docs/painless/painless-operators-boolean.asciidoc]: Maybe, "Using the greater than operator with non-numeric expressions results in an error."?

File: docs/painless/painless-operators-boolean.asciidoc [hunk 4]
```diff
@@ -0,0 +1,516 @@
...
+Greater than or equal compares two numerical expressions where a resultant boolean value will be true if the left-side expression is a larger value than or equal to the right-side expression otherwise false.  The format is an expression, followed by the right angle and equals operator, and a closing expression.
+
+*Grammar:*
+[source,ANTLR4]
+----
+greater_than_or_equal: expression '>=' expression;
+----
+
+A numeric promotion may occur during a greater than or equal operation.  A def type evaluated at run-time will follow the same promotion table at run-time following whatever type def represents. Non-numeric expressions will result in an error.
+
```
Comment (debadair) [inline docs/painless/painless-operators-boolean.asciidoc]: Using the greater than or equal operator with non-numeric expressions results in an error. (Same for following sections)

File: docs/painless/painless-operators-boolean.asciidoc [hunk 5]
```diff
@@ -0,0 +1,516 @@
...
+Less than compares two numerical expressions where a resultant boolean value will be true if the left-side expression is a smaller value than the right-side expression otherwise false.  The format is an expression, followed by the left angle operator, and a closing expression.
+
+*Grammar:*
+[source,ANTLR4]
+----
+less_than: expression '<' expression;
+----
+
+A numeric promotion may occur during a less than operation.  A def type evaluated at run-time will follow the same promotion table at run-time following whatever type def represents. Non-numeric expressions will result in an error.
+
```
Comment (debadair) [inline docs/painless/painless-operators-boolean.asciidoc]: Same as previous.

File: docs/painless/painless-operators-boolean.asciidoc [hunk 6]
```diff
@@ -0,0 +1,516 @@
...
+boolean x = false;    // declares the boolean variable x and sets the constant boolean false
+boolean y = x ^ true; // declares the boolean variable y and sets it the result of x xor true
+def z = y ^ x;        // declares the def variable z and sets it to the result of y xor x
+----
+
+[[boolean-and-operator]]
+==== Boolean And
+
+Boolean and will and together two boolean expressions.  If the first expression is found to be false then it is known that the result will also be false, so evaluation of the second expression will be skipped.  The table below shows what the resultant boolean value will be based on the two boolean expressions.
+
```
Comment (debadair) [inline docs/painless/painless-operators-boolean.asciidoc]: I'd probably go with, "Boolean `and` ANDs two boolean expressions." Is it worth adding, "and evaluates to true if both expressions are true"?

(Also corresponding changes for the following sections.)

File: docs/painless/painless-operators-general.asciidoc [hunk 7]
```diff
@@ -0,0 +1,426 @@
+[[painless-operators-general]]
+=== Operators: General
+
+[[precedence-operator]]
+==== Precedence
+
+An expression encapsulated within a *precedence operator* overrides
+existing precedence relationships between operators and is evaluated
```
Comment (debadair) [inline docs/painless/painless-operators-general.asciidoc]: Need to call out the use of parens. The use of "precedence operator" seems a bit awkward in this case--and doesn't need to be highlighted, especially on the second occurrence. I might be inclined to go with something like, "An expression encapsulated by the precedence operator (enclosed in parentheses)..."

File: docs/painless/painless-operators-numeric.asciidoc [hunk 8]
```diff
@@ -0,0 +1,663 @@
+[[painless-operators-numeric]]
+=== Operators: Numeric
+
+[[post-increment-operator]]
+==== Post Increment
+
+A variable/field representing a numerical value can be possibly evaluated as part of an expression, and then increased by 1 for its respective type.  The format starts with a variable name followed by a plus and ends with a plus.
+
```
Comment (debadair) [inline docs/painless/painless-operators-numeric.asciidoc]: Should follow the same format used for the other operators--something like, "use the post increment operator (++) to increase the value of a variable by one after it has been evaluated. The format is the variable name followed by `++`.

File: docs/painless/painless-operators-numeric.asciidoc [hunk 9]
```diff
@@ -0,0 +1,663 @@
...
+float l = 1.0f; // declares the float variable l and sets it the constant float 1.0f
+float k;        // declares the float variable k
+k = l--;        // sets the float variable k to the value of l (1.0f), and then decrements the float variable l by 1.0 to a value of 0.0
+----
+
+[[pre-increment-operator]]
+==== Pre Increment
+
+A variable/field representing a numerical value can be increased by 1 for its respective type, and then possibly evaluated as part of an expression.  The format starts with a plus followed by a plus and ends with a variable name.
+
```
Comment (debadair) [inline docs/painless/painless-operators-numeric.asciidoc]: See prev comment.

**Review by debadair — COMMENTED** [review]
Left some comments--where applicable, the changes can be carried forward to the other sections, I didn't repeat them in each case.

File: docs/painless/painless-casting.asciidoc [hunk 10]
```diff
@@ -438,19 +441,82 @@ type value is derived from what the `def` type value represents at run-time.
...
+    store `float 3.0` to `f`;
     (note this example illustrates promotion done at run-time as promotion
             done at compile-time would have resolved to a `def` type value)
+
+[[allowed-casts]]
+==== Allowed Casts
+
+The following tables list all allowed casts read as row (original type) to
+column (target type) with implicit as `I`, explicit as `E`, and not allowed as
+`-`:
```
[the remark this answers is not in the document]
Comment (jdconrad) [reply docs/painless/painless-casting.asciidoc]: Changed.

File: docs/painless/painless-literals.asciidoc [hunk 11]
```diff
@@ -44,7 +44,7 @@ HEX:     '-'? '0' [xX] [0-9a-fA-F]+ [lL]?;
 <5> `int -18` in octal
 <6> `int 3882` in hex
 
-[[floats]]
+[[float-literals]]
 ==== Floats
```
[the remark this answers is not in the document]
Comment (jdconrad) [reply docs/painless/painless-literals.asciidoc]: Thanks for the head up.  I've been pretty good about compiling the docs and looking at the format each time I change something significant.

File: docs/painless/painless-operators-general.asciidoc (hunk 7 above)
Comment (jdconrad) [reply docs/painless/painless-operators-general.asciidoc]: Cleaned up.  I removed the format sentence from each operator given the grammar and examples.  Is this something I should add back in for each operator?

Comment (jdconrad) [reply docs/painless/painless-operators-general.asciidoc]: Also removed the bolding of the operators from all of the general section.

File: docs/painless/painless-operators-array.asciidoc [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+
+[[array-initialization-operator]]
+==== Array Initialization
+
+You create and initialize arrays using the brackets `[]` and braces `{}`
+operators. Each set of brackets represents a dimension. The values you want to
+initialize each dimension with are specified as a comma-separated list enclosed
+in braces. For example, `new int[] {1, 2, 3}` creates a one dimensional `int`
+array with a size of 3 and the values 1, 2, and 3.
+
```
[the remark this answers is not in the document]
Comment (jdconrad) [reply docs/painless/painless-operators-array.asciidoc]: Changed.

File: docs/painless/painless-operators-boolean.asciidoc (hunk 1 above)
Comment (jdconrad) [reply docs/painless/painless-operators-boolean.asciidoc]: Used caps, tried to do this consistently for all operators that it made sense for.

File: docs/painless/painless-operators-boolean.asciidoc (hunk 2 above)
Comment (jdconrad) [reply docs/painless/painless-operators-boolean.asciidoc]: Changed.

File: docs/painless/painless-operators-boolean.asciidoc (hunk 3 above)
Comment (jdconrad) [reply docs/painless/painless-operators-boolean.asciidoc]: Changed.

File: docs/painless/painless-operators-boolean.asciidoc (hunk 4 above)
Comment (jdconrad) [reply docs/painless/painless-operators-boolean.asciidoc]: Changed.

File: docs/painless/painless-operators-boolean.asciidoc (hunk 5 above)
Comment (jdconrad) [reply docs/painless/painless-operators-boolean.asciidoc]: Changed.

File: docs/painless/painless-operators-boolean.asciidoc (hunk 6 above)
Comment (jdconrad) [reply docs/painless/painless-operators-boolean.asciidoc]: Changed, used caps.

File: docs/painless/painless-operators-numeric.asciidoc (hunk 8 above)
Comment (jdconrad) [reply docs/painless/painless-operators-numeric.asciidoc]: Changed.

File: docs/painless/painless-operators-numeric.asciidoc (hunk 9 above)
Comment (jdconrad) [reply docs/painless/painless-operators-numeric.asciidoc]: Changed.

**Comment by jdconrad** [comment]
@debadair Thank you so much for the first pass!

As we discussed, I've merged ALL changes to the operators section into this PR.  I apologize profusely that it's so huge, I realize the most that can be done as a second pass is a broad overview and will need editing again when you have more time, but I think it makes sense to get these changes in as I consider them an large improvement over the existing docs.

I did my best to apply comments from the previous pass as appropriate to the different operators, so hopefully it's better moving forward.

Note - I also merged in more structural changes as I figured I may as well try to lay out the structure for the spec all at once so you don't have to request so many hyperlinks in upcoming new documentation.

The sections are the following:

* painless-lang-spec.asciidoc
** painless-comments.asciidoc
**painless-keywords.asciidoc
**painless-literals.asciidoc
**painless-identifiers.asciidoc
**painless-variables.asciidoc
**painless-types.asciidoc
**painless-casting.asciidoc
**painless-operators.asciidoc
**painless-operators-general.asciidoc
**painless-operators-numeric.asciidoc
**painless-operators-boolean.asciidoc
**painless-operators-reference.asciidoc
**painless-operators-array.asciidoc
**painless-statements.asciidoc
**painless-scripts.asciidoc
**painless-functions.asciidoc
**painless-lambdas.asciidoc
**painless-regexes.asciidoc

File: docs/painless/painless-casting.asciidoc [hunk 13]
```diff
@@ -343,12 +347,12 @@ reference type to its corresponding primitive type.
 
 Implicit boxing/unboxing occurs during the following operations:
 
-* Conversions between a `def` type and a primitive type will be implicitly
+* Conversions between a `def` type and a primitive type is implicitly
   boxed/unboxed as necessary, though this is referred to as an implicit cast
```
Comment (debadair) [inline docs/painless/painless-casting.asciidoc]: is > are

File: docs/painless/painless-literals.asciidoc [hunk 14]
```diff
@@ -114,9 +117,9 @@ STRING: ( '"'  ( '\\"'  | '\\\\' | ~[\\"] )*? '"'  )
 "double-quoted with non-escaped 'single-quotes'"
 ----
 
-[[characters]]
+[[character-literals]]
 ==== Characters
 
-A character literal cannot be specified directly. Instead, use the
+A character literal is not specified directly. Instead, use the
 <<string-character-casting, cast operator>> to convert a `String` type value
```
Comment (debadair) [inline docs/painless/painless-literals.asciidoc]: I would change this to "Character literals are not..." It just reads strange to me in the singular. :-)

File: docs/painless/painless-scripts.asciidoc [hunk 15]
```diff
@@ -0,0 +1,6 @@
+[[painless-scripts]]
+=== Scripts
+
+Scripts are composed of one-to-many <<painless-statements, statements>> and are
+run in a sandbox that determines what local variables are immediately available
+along with what API's are whitelisted for use.
```
Comment (debadair) [inline docs/painless/painless-scripts.asciidoc]: No apostrophe necessary in APIs

**Review by debadair — COMMENTED** [review]
A few more very minor comments. This is definitely  feeling like a big improvement--nice work!  Totally agree with getting this merged!

File: docs/painless/painless-scripts.asciidoc (hunk 15 above)
Comment (jdconrad) [reply docs/painless/painless-scripts.asciidoc]: Fixed.

File: docs/painless/painless-literals.asciidoc (hunk 14 above)
Comment (jdconrad) [reply docs/painless/painless-literals.asciidoc]: Agreed and changed.

File: docs/painless/painless-casting.asciidoc (hunk 13 above)
Comment (jdconrad) [reply docs/painless/painless-casting.asciidoc]: Fixed.

**Comment by jdconrad** [comment]
@debadair Thanks again for the feedback!  Made the changes as requested.

**Review by debadair — APPROVED** [review]

**Comment by jdconrad** [comment]
@debadair Thank you so much!  Will commit this as soon as the build passes.  I will try to keep the next rounds much, much smaller.

**Comment by jdconrad** [comment]
@elasticmachine Please test this.

**Pull request merged**